### ISSUE TITLE
The Complete Bus Logic of the Intel 8088

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#10889]
The Complete Bus Logic of the Intel 8088<br>
<br>
https://ift.tt/NGMKs7m<br>
<br>
<br>
<br>

<div><div><div><div>
<div>
<p>In a <a href="https://martypc.blogspot.com/2023/08/the-8088-prefetch-algorithm.html" target="_blank">previous article</a>, I described a simple state machine to explain the prefetch behavior and miscellaneous observed bus delays of the Intel 8088. I now believe that the model I described there - although successful in modelling the 8088's behavior - was unnecessarily complex. There are in fact far fewer legitimate bus "delays", and they do not relate to internal state transitions. I have also slightly revised my description of the BIU's prefetching logic - I now believe that the prefetching decision happens at the end of T2, instead of the beginning of T3. There may not be much of a distinction between the end of one cycle and the beginning of the next, but it lines up better in our new diagrams.</p>
<h2>Introduction</h2>
<p>If you were interested enough to click on this article, you are probably already familiar with the 8088's bus cycle timings. A single bus cycle on the 8088 takes a minimum of 4 CPU cycles, consisting of 4 T-states numbered T1, T2, T3 and T4. Wait states (Tw) can be inserted between T3 and T4 via manipulation of the CPU's READY line to accommodate access to slower devices, but in general, the 4 T-states have been the demonstrable model of the 8088 for decades.  </p>
<p></p>
<table cellspacing="0" align="center" cellpadding="0"><tbody>
<tr><td><a href="https://blogger.googleusercontent.com/img/b/R29vZ2xl/AVvXsEhOceW4Sxc2KnRrGvt_Q9_w908BZ680kiykVlXgiERdjH5-7G1_jGh0ra6Rz7vxD-ECQh23K5diZYCq9MTdc4ecPA5sBHaOrTHmzbsa3mLqYC5KnQ3FP6aqwc-hGg0p6a2jDwX4dJYf-Tb7tzJhY2urTDsQp-EXe7KeoX1utl4U08bzhLZeOW3Fv8Q-b7aX/s1098/bus_cycle-4-8_01.png"><img border="0" src="https://blogger.googleusercontent.com/img/b/R29vZ2xl/AVvXsEhOceW4Sxc2KnRrGvt_Q9_w908BZ680kiykVlXgiERdjH5-7G1_jGh0ra6Rz7vxD-ECQh23K5diZYCq9MTdc4ecPA5sBHaOrTHmzbsa3mLqYC5KnQ3FP6aqwc-hGg0p6a2jDwX4dJYf-Tb7tzJhY2urTDsQp-EXe7KeoX1utl4U08bzhLZeOW3Fv8Q-b7aX/w640-h96/bus_cycle-4-8_01.png"></a></td></tr>
<tr><td><p><i>iAPX 86/88 User's Manual, pg 4-5</i></p></td></tr>
</tbody></table>
<p></p>
<p></p>
<p>Diagrams such as the above have appeared in innumerable manuals, web references and computer science textbooks. It is no surprise then that emulator authors attempted to construct the logic and timings of their emulators around a 4-cycle bus cycle time. Unfortunately, this is an incomplete picture. For whatever reason, Intel was averse to advertising the true bus cycle time of the 8088 in their official documentation, and so this incomplete model has persisted to this day.</p>
<p>Ken Shirriff, in his analysis of the 8086's silicon, was the first to document this, to my knowledge, on his blog entry, <a href="https://www.righto.com/2023/07/8086-pins.html" target="_blank">Reverse-engineering the 8086 processor's address and data pin circuits</a>.</p>
<p>He states,</p>
<p><span></span></p>
<blockquote><b>There [are] two undocumented T states that are important to the bus cycle. The physical address is computed in the two clock cycles before T1 so the address will be available in T1. I give these "invisible" T states the names TS (start) and T0.</b></blockquote>
<p>I originally missed the significance of this - it's casually mentioned as if it were an interesting, but ultimately trivial internal detail. However these cycles are fundamental to the timing and operation of the 8088. They are not quite "invisible" after all!</p>
<h1><span>To properly emulate the 8088, you must account for at least a 6 cycle bus cycle time.</span></h1>
<p></p>
<p>It seems farfetched at first - we know that we can execute NOPs back to back at the speed at which the opcodes can be fetched - that is 4 cycles for each byte-sized NOP opcode. So that would disprove this theory right out of the gate.  The key that makes it possible is that the BIU is actually <i>pipelined. </i></p>
<p>The Intel iAPX manual gave us a hint, even if it only mentioned the four conventional T-states:</p>
<p>A <i>pipelined architecture</i>.  The only previously pipelining I was really aware of was the ability of the 8088 to execute the last cycle of an instruction on the first cycle of a new instruction, via the NXT microcode flag facility. But it turns out the 8088 is capable of pipelining the calculation of the 20-bit address for the next bus cycle while it is performing the first.</p>
<p>In retrospect, the need for additional bus cycles should have been obvious. Consider what happens on the T1 state of the 8088.  On T1, the full, 20-bit address of the bus cycle operation is placed on the CPU's address pins, and the ALE signal - either directly or via the i8288, is asserted so that the address can be latched by motherboard circuitry.</p>
<p><br></p>
<p>Remember that this is a CPU that takes 3 cycles to add an immediate operand to a register.  While it's true that the BIU has a dedicated adder to produce the 20 bit address (seen above on the right hand side at the top), before this adder can produce the address emitted on T1 it must first load the correct segment register and index register, and only then perform the shift and addition operation. </p>
<p><br></p>
<p>This takes two cycles, which Ken has named Ts and T0.</p>
<p><br></p>
<div><p>This makes it impossible that any bus request from the EU can result in a T1 immediately, even if the prefetcher is idle. Observationally, after mining a <a href="https://martypc.blogspot.com/2023/09/a-test-suite-for-intel-8088.html" target="_blank">dataset of over 3 million instruction executions</a>, I found that the minimum delay between a EU bus request and the subsequent T1 is 3 cycles. This is just not usually noticed because much of the time during normal operation the EU is contending with the BIU's prefetch unit for the bus.  One might naively assume that, if it were not for prefetch bus cycles getting in the way, an EU bus request might have been serviced faster - even immediately.</p></div>
<p><br></p>
<p>I am adding a third logical T-state to the two additional states already mentioned: Tr.  This may or may not correspond to a physical state in silicon, but it represents the cycle on which either a prefetch decision or bus request from the EU is made. Therefore in the model I am about to describe, the total bus cycle time of the 8088 can be thought of as <b>7</b> cycles in total!</p>
<p><br></p>
<p>As a base example, lets look at our NOPs executing back to back. First, we will present the traditional model of a 4-cycle bus cycle time:</p>
<p><br></p>
<table cellspacing="0" align="center" cellpadding="0"><tbody>
<tr><td><a href="https://blogger.googleusercontent.com/img/b/R29vZ2xl/AVvXsEjLVOLHhOgaT3YRNHFAQbnj3AdthaceN_ookz-iefzKNb8xMdxRdFzGg4BLckmvR6VVD4SXz7v8GPNK-gCfPHfLzcSsEZRpK-PcGOr0FxGizy7o24fRG8pSfxN2OBgqbgApk7uRaFjtDjz1Hhl8Gd2KvBuVxr-QqB_kQvyijlTmnECE5Xxm5qbfDcstH8H5/s767/4_cycle_nops_01.png"><img border="0" src="https://blogger.googleusercontent.com/img/b/R29vZ2xl/AVvXsEjLVOLHhOgaT3YRNHFAQbnj3AdthaceN_ookz-iefzKNb8xMdxRdFzGg4BLckmvR6VVD4SXz7v8GPNK-gCfPHfLzcSsEZRpK-PcGOr0FxGizy7o24fRG8pSfxN2OBgqbgApk7uRaFjtDjz1Hhl8Gd2KvBuVxr-QqB_kQvyijlTmnECE5Xxm5qbfDcstH8H5/s16000/4_cycle_nops_01.png"></a></td></tr>
<tr><td>The traditional, incomplete bus cycle model</td></tr>
</tbody></table>
<p><br></p>
<p>Here we see NOPs executing back to back while the queue is empty. Although NOP is a 3-cycle instruction, they are lengthened here to four due to needing to wait an additional cycle for the the next NOP opcode to be fetched and read from the queue.  We cannot read bytes out of an empty queue any faster than they are put in, so our NOPs all execute in four cycles. There's no evidence here of a longer bus cycle time than 4.</p>
<p><br></p>
<p>Let's look at an alternative diagram with the additional, pipelined T-states illustrated:</p>
<p><br></p>
<table cellspacing="0" align="center" cellpadding="0"><tbody>
<tr><td><a href="https://blogger.googleusercontent.com/img/b/R29vZ2xl/AVvXsEjCPbAxYZOXB5MMBPzu5qgX8f0zrM797QQA93nuxZlwo-14tlypM0TtfOYHePZfFX5AK6Bhtb_1aztunJ53dEnkgW7-wu_KaEYU70Sphai9HH2kgbnOGbsBkngKGyaTi6KUoj1QqRhBFxkk-Vpl2LOcvrFG6XPj7nifcKcN9q6yLG5_IInKk23rnIFRbhDv/s791/7_cycle_nops_01.png"><img border="0" src="https://blogger.googleusercontent.com/img/b/R29vZ2xl/AVvXsEjCPbAxYZOXB5MMBPzu5qgX8f0zrM797QQA93nuxZlwo-14tlypM0TtfOYHePZfFX5AK6Bhtb_1aztunJ53dEnkgW7-wu_KaEYU70Sphai9HH2kgbnOGbsBkngKGyaTi6KUoj1QqRhBFxkk-Vpl2LOcvrFG6XPj7nifcKcN9q6yLG5_IInKk23rnIFRbhDv/s16000/7_cycle_nops_01.png"></a></td></tr>
<tr><td>Our new logical 7-cycle bus cycle model</td></tr>
</tbody></table>
<p>At the end of T2, a prefetch decision is made.  In this case, any possible Tr state is fleeting at best. Its entirely possible that Tr is only really meaningful for EU bus requests, but I include it here for consistency's sake, as it shows the cycle on which the prefetch decision is made. The bus cycle for the next fetch begins with the Ts state on the previous cycle's T3, so that a T1 can immediately occur after T4.</p>
<p><br></p>
<h2>BIU Prefetch Logic</h2>
<p>Let's take a moment to recap the logic of prefetching.</p>
<p><br></p>
<p>Prefetching is an automatic process performed by the BIU, where bytes from the code segment are read from memory and placed into the processor's instruction queue. This process happens proactively as long as prefetching is not suspended and there is room in the queue to place a new byte.</p>
<p><br></p>
<p>Prefetching can be suspended in certain scenarios:</p>
<div><ul>
<li>During the <b>HALT </b>state</li>
<li>After execution of the <b>SUSPEND </b>microcode operation</li>
<li>When the queue is full</li>
</ul></div>
<p>A prefetch is typically started* at the end of T2 during the last byte of any atomic bus transfer. On the 8088, that is an important point, since word transfers are split into two byte transfers back to back.  During the bus cycle in which the first byte of a word is being transferred, a prefetch will not be started. </p>
<p><br></p>
<blockquote><p>*My previous article discussed prefetches as being "scheduled" - implying some latency between the decision to fetch and the initiation of the code fetch bus cycle. Now that we have a model of the full bus cycle time, there is actually no latency at all.  Thus a "scheduled prefetch" isn't a concept we need. A fetch is simply started on a certain cycle or it is not.</p></blockquote>
<p>A prefetch can be started in a few other scenarios:</p>
<p></p>
<ul>
<li>A <b>FLUSH </b>microcode operation is executed, resuming fetching from a <b>SUSPEND</b>ed state.</li>
<li>A byte is read from a full queue, making room for a new byte</li>
<li>The bus is otherwise idle, and there is room for the queue for a byte to be fetched</li>
</ul>
<p>A fetch can also be delayed, as a policy decision, at the end of T2 when the length of the queue is 3.  This effectively represents a 3 cycle delay, but specific numbers need not be modelled - if we simply skip the prefetch decision on T2 when the queue is 3, then the fetch will be started on the cycle after T4, implementing our 3 cycle delay with no extra logic required. We will discuss this more in detail later.</p>
<p><br></p>
<p>A pending request from the EU on or prior to T2 will prevent a prefetch from beginning at the end of T2. </p>
<p><br></p>
<p>If a request from the EU is received on T3 or T4 after a prefetch bus cycle has commenced, the prefetch can be <b>aborted </b>on T4.  This traditionally required modelling two Ti idle states known as a <b>prefetch abort penalty</b>. As we will see with our new model, no delays or penalty cycles are actually required.</p>
<h3>Bootstrapping Prefetch</h3>
<p>We can observe that a code fetch T1 cycle begins immediately after the 8088 completes the <b>RESET </b>microcode subroutine from power-on.  Our previous model had to describe a state transition to explain the delay between the <b>FLUSH </b>command that initiates the CPU's first fetch and the actual <b>T1</b> of the fetch:</p>
<p><br></p>
<table cellspacing="0" align="center" cellpadding="0"><tbody>
<tr><td><a href="https://blogger.googleusercontent.com/img/b/R29vZ2xl/AVvXsEgk9k42_xP8BYo3j_mAWr5cyPd-8msvlcDuzDyRumn-joDN8jEudL3ARvKCd3n9lIjKNNLV5dqOk30hC0H8GNfbtQjduTDCyGPUZtnMLk7rGcwlvef42zhXzhG9xRWeWdP1PsuSdqa8d5XMNjnvvpjrm5zB4wsa1xdCylin5NpJH0fc4m8k3vQg5pXafKwY/s768/fetch_boostrap_old_01.png"><img border="0" src="https://blogger.googleusercontent.com/img/b/R29vZ2xl/AVvXsEgk9k42_xP8BYo3j_mAWr5cyPd-8msvlcDuzDyRumn-joDN8jEudL3ARvKCd3n9lIjKNNLV5dqOk30hC0H8GNfbtQjduTDCyGPUZtnMLk7rGcwlvef42zhXzhG9xRWeWdP1PsuSdqa8d5XMNjnvvpjrm5zB4wsa1xdCylin5NpJH0fc4m8k3vQg5pXafKwY/s16000/fetch_boostrap_old_01.png"></a></td></tr>
<tr><td>RESET routine, old bus model</td></tr>
</tbody></table>
<p>In the diagrams I will be showing, a pink color indicates previously unexplained or state-modelled delays that are no longer needed.</p>
<p><br></p>
<p>If we simply enter the Tr state on <b>FLUSH</b>, then everything follows naturally, with no artificial delays:</p>
<p><br></p>
<table cellspacing="0" align="center" cellpadding="0"><tbody>
<tr><td><a href="https://blogger.googleusercontent.com/img/b/R29vZ2xl/AVvXsEjcj0lmP_wxlDWXI_R4WgpwdUyncXbwlYbix3sp6GxmQ-LyHLLXIQs0XqvAjJjBz3U0Kes2P65xkY9DnNg7s6V6Bulk1sE7SyV8aznyfm0QQyfNB9dHScvD99mGqK2WJtpMzqpav7bHFmwXAPUETzO9pKNFKxjpau6NdvUkoraS8ipJOVBSlmzm8J58Oh0Q/s794/fetch_bootstrap_01.PNG"><img border="0" src="https://blogger.googleusercontent.com/img/b/R29vZ2xl/AVvXsEjcj0lmP_wxlDWXI_R4WgpwdUyncXbwlYbix3sp6GxmQ-LyHLLXIQs0XqvAjJjBz3U0Kes2P65xkY9DnNg7s6V6Bulk1sE7SyV8aznyfm0QQyfNB9dHScvD99mGqK2WJtpMzqpav7bHFmwXAPUETzO9pKNFKxjpau6NdvUkoraS8ipJOVBSlmzm8J58Oh0Q/s16000/fetch_bootstrap_01.PNG"></a></td></tr>
<tr><td>RESET routine, new bus model</td></tr>
</tbody></table>
<h3>Prefetch Aborts</h3>
<p>Since a prefetch is normally started at the end of T2 of a bus cycle, a bus request from the EU prior to or on T1 or T2 will enter the Tr state and prevent a fetch from beginning during that bus cycle.</p>
<p><br></p>
<p>At the end of T2, a fetch bus cycle is started, and we enter the T3-Ts pipelined state.  If an EU bus request is received now either on T3-Ts, or T4-T0, then the prefetch bus cycle can be cancelled, and a new bus cycle servicing the EU immediately begun.</p>
<p><br></p>
<p>Here is what this looks like under the traditional bus model:</p>
<table cellspacing="0" align="center" cellpadding="0"><tbody>
<tr><td><a href="https://blogger.googleusercontent.com/img/b/R29vZ2xl/AVvXsEg1l7sSqhpo_Uawn8ckDcuBVSdebCe5eD_Ij31ei69cWROsnDpOlx0rlSlJKUMuArpRFuoupbtGLrfcCdHHzxHlKfo81klgAnZbbHGAqvb3HWNlRXB9JWyi5-2wRWwqo238X7l28DP-v7mqzScqV5OFupgxASQqvwp7RY6Fmurp8rGxoLdRBncqvYYWR-1k/s769/fetch_abort_old_01.png"><img border="0" src="https://blogger.googleusercontent.com/img/b/R29vZ2xl/AVvXsEg1l7sSqhpo_Uawn8ckDcuBVSdebCe5eD_Ij31ei69cWROsnDpOlx0rlSlJKUMuArpRFuoupbtGLrfcCdHHzxHlKfo81klgAnZbbHGAqvb3HWNlRXB9JWyi5-2wRWwqo238X7l28DP-v7mqzScqV5OFupgxASQqvwp7RY6Fmurp8rGxoLdRBncqvYYWR-1k/s16000/fetch_abort_old_01.png"></a></td></tr>
<tr><td>The traditional 4-cycle bus model with its "prefetch abort" penalty cycles</td></tr>
</tbody></table>
<p><br></p>
<p>The picture is much simpler in the new model:</p>
<p><br></p>
<p>We can see that on T4, the EU bus request is immediately serviced. The previous cycles we explained as "penalty cycles" are simply the required Ts and T0 states to recalculate the new address for the EU bus cycle, with the work of the previous Ts cycle intended for a fetch discarded.</p>
<p><br></p>
<p>There <i>is</i> an effective delay here, in that the abort check only occurs on T4. Therefore if we moved the bus request from the EU one cycle earlier to T3, the cycle timings illustrated would be identical, with Tr held for two cycles. </p>
<p><br></p>
<p>This logic finally explains a paragraph in the Intel iAPX 86 88 User's Guide which had confused me since I first read it:</p>
<p><br></p>
<table cellspacing="0" align="center" cellpadding="0"><tbody>
<tr><td><a href="https://blogger.googleusercontent.com/img/b/R29vZ2xl/AVvXsEjhxRM0aFq0ycl6ynZ2ulmH99svSVy8zx3mNiam9NWjHXDa2eUDMRrS_TnD0ct7bHE3tzi7FgolX_6A2rMofoaF9Fb-ZGEHwXHKfuhALjljOSP1xWCfFLLjjuH9MQt1Epm2mQLV90CECZY90exDdrqreV68TL0QGnVy27h7sxZs5C6Nrkb4ddUQHpcHj2G0/s767/delay_paragraph_01.PNG"><img border="0" src="https://blogger.googleusercontent.com/img/b/R29vZ2xl/AVvXsEjhxRM0aFq0ycl6ynZ2ulmH99svSVy8zx3mNiam9NWjHXDa2eUDMRrS_TnD0ct7bHE3tzi7FgolX_6A2rMofoaF9Fb-ZGEHwXHKfuhALjljOSP1xWCfFLLjjuH9MQt1Epm2mQLV90CECZY90exDdrqreV68TL0QGnVy27h7sxZs5C6Nrkb4ddUQHpcHj2G0/w400-h175/delay_paragraph_01.PNG"></a></td></tr>
<tr><td><i>iAPX 86/88 User's Manual, 1-109</i></td></tr>
</tbody></table>
<p>This paragraph, when read with the mental model of a 4-cycle bus cycle time, seems to contradict itself. It states that a prefetch abort results in two idle cycles, and yet, goes on to state that the maximum delay incurred is one cycle. Well, which is it? Is the delay two cycles, or one?</p>
<p><br></p>
<p>It's clear now - the maximum delay incurred is one cycle in the case where the EU request arrives on T3 and must wait for the abort decision on T4. But the two 'idle' Ti cycles mentioned are not really idle at all - they are the Ts and T0 states. Intel must have had their reasons for not wanting to disclose the reality of their 6-cycle bus access time, but left this passage as a clue.</p>
<p><br></p>
<h3>Wait States</h3>
<p>The CPU will normally attempt to perform the actual bus cycle operation, a read or a write, on T3 of the bus cycle. However, the addressed device can signal that is needs more time to respond via the CPU's READY line. When this occurs, wait states, or Tw cycles, are inserted between T3 and T4. This could conceivably have an affect on our bus cycles are pipelined.</p>
<p><br></p>
<p>In this example, 3 wait states have been injected into memory access, affecting the the fetch that occurs prior to the requested IO write operation. </p>
<p><br></p>
<p>It should be clear that a T0 state cannot proceed to T1 until T4 has completed - the BIU can simply hold on to the computed address for the next bus cycle, repeating T0 until T1. Likewise, since a prefetch abort decision happens on T4, the wait states add to the effective latency of the EU bus request made on the first Tw state. There is effectively now a four-cycle window in which the EU could have made its request without affecting overall cycle execution time of this instruction.</p>
<p><br></p>
<h2>Explaining the 8088 Bus Delays</h2>
<p>With a 4-cycle bus cycle model, the 8088 encounters confusing delays at times that cannot be accounted for by microcode execution or an active bus transfer.  Convoluted logic is required to handle them.  Properly modelling the a 7-cycle bus cycle time explains these delays simply in every circumstance I have encountered. Here we will examine a few of them.</p>
<h3>SUSPEND "Delays"</h3>
<p>In a 4-cycle bus cycle model, it can be observed that when prefetching is disabled, either via a full queue, or via the <b>SUSPEND </b>microcode routine, that EU bus requests appear delayed.  These delays required contrived state logic to explain previously. Now, they are simply the result of the full 7-cycle bus cycle time being exposed since there are no fetches to pipeline the first 3 cycles in. </p>
<p><br></p>
<p>The <b>CALL FAR</b> instruction is interesting in that it issues the <b>SUSPEND </b>microcode routine, but then issues bus requests while the prefetcher is suspended.  On the left, we see the mysterious delays in pink. On the right, we simply see the true, 7-cycle bus cycle timings, with no delay states or state transitions required.</p>
<p><br></p>
<p><br></p>
<table cellspacing="0" align="center" cellpadding="0"><tbody>
<tr><td><a href="https://blogger.googleusercontent.com/img/b/R29vZ2xl/AVvXsEiIGs5SZaz1RiblzuKDwGKOp6xDbQXM3FdroPbGGQoMLIoDUJnORE0hTGWJ0rF30gAu_lmcopbKHmO0GvwFTRzlGkE9CW13GhQ1_QbfMbze5NgyftWpxCTqGnb07cx3TH2eydnKBI9pGi2HhzXsfPGVekRSjpArniTDUD4CoBJ0_mAu6nkVwOGlbbCTwysg/s1335/request_while_suspend_01.png"><img border="0" src="https://blogger.googleusercontent.com/img/b/R29vZ2xl/AVvXsEiIGs5SZaz1RiblzuKDwGKOp6xDbQXM3FdroPbGGQoMLIoDUJnORE0hTGWJ0rF30gAu_lmcopbKHmO0GvwFTRzlGkE9CW13GhQ1_QbfMbze5NgyftWpxCTqGnb07cx3TH2eydnKBI9pGi2HhzXsfPGVekRSjpArniTDUD4CoBJ0_mAu6nkVwOGlbbCTwysg/s16000/request_while_suspend_01.png"></a></td></tr>
<tr><td>Old vs New Bus Model</td></tr>
</tbody></table>
<p><br></p>
<h3>Queue Full "Delays"</h3>
<p>A similar phenomenon occurs when the instruction queue is full, which affects long running instructions that make memory accesses, such as string operations like <b>REP MOVSB.  </b>Every read and write that the instruction makes while the queue is full (and thus prefetching suspended) appears to be delayed by 3 cycles. </p>
<p><br></p>
<p>Under our new model, these delays vanish, each bus request entering the Tr state immediately when requested:</p>
<p><br></p>
<p><br></p>
<p><br></p>
<p>While we are on the topic of <b>MOVS </b>instructions: On the 8088, <b>REP MOVSW</b> is much faster than <b>REP MOVSB</b>, despite the 8088 being limited to an 8 bit data bus, which one might assume would preclude any major advantage of one over the other. Part of that is due to halved overhead in the microcode loop logic, but that is still only 3 cycles per iteration - loop overhead alone does not account for the difference. </p>
<p><br></p>
<p>If we look at the published cycle times for <b>REP MOVSB</b>, we can see that it listed as 17 cycles per iteration.  With 3 cycles of microcode loop overhead, that leaves 14 cycles to actually move two bytes - matching our new model of 7 cycles per bus cycle.  A word transfer, however, can take advantage of the BIU pipeline for the second byte, completing in only 11 cycles.   11+11+3 = 25 cycles per iteration, again matching published timings. Therefore, a good percentage of the speed advantage of <b>MOVSW </b>on the 8088 is due to bus pipelining.</p>
<p><br></p>
<h3>Prefetch Policy Delays</h3>
<p>During a code fetch cycle, at the end of T2 when a prefetch decision is made, the length of the queue is examined. If the queue length is 3, a prefetch is not started.</p>
<p><br></p>
<p>Under the old bus model, this required modelling a 3-cycle prefetch delay:</p>
<p>Under our new model, we simply do not begin a fetch on T2 when the queue length is 3. After T4, a fetch will begin if the queue is no longer full.</p>
<p><br></p>
<p>Our three cycle 'delay' is now simply the exposed, full 7-cycle bus cycle. </p>
<p></p>
<p>If we did not have this special policy, were in an active code fetch bus cycle, and the queue length at the end of T2 was 3, then it can be assumed that - barring intervention from the EU - on T4 the queue would be filled by the current code fetch.  But we would have already begun a new code fetch bus cycle - without any room to put the byte we intend to fetch!</p>
<p><br></p>
<p>Thus this policy decision prevents a scenario where a code fetch cycle is begun where there is a possibility that it could not complete. After T4, we know whether we can safely fetch again.  For this reason, the policy decision is only applied during a code fetch cycle.</p>
<h2>Interrupt Timing</h2>
<p>Interrupt handling is a microcoded subroutine - just like any instruction, the request for an Interrupt Acknowledge bus cycle is issued from the EU, and so, interrupts follow the same rules.  However, there is additional latency incurred by loading and beginning the interrupt microcode routine.</p>
<p><br></p>
<p><br></p>
<table cellspacing="0" align="center" cellpadding="0"><tbody>
<tr><td><a href="https://blogger.googleusercontent.com/img/b/R29vZ2xl/AVvXsEhm-HVxRc2MEIpbNis7OoTBLqfWnsiy-iNd3I6EaAcc3wlyVzDwxnpLmoTIA0gi8I6yxVHgSkWxFXndvWivWDMn6YFAmF6TS-7hc2yGIhZRWunNfynVlwspvfYeUH5-zYzV7f0VGimOcDI7W3YtUEUYHZQU0r8kY_NprDVTamIOD600PBohHIb3cxv4t-bl/s796/interrupt_new_01.png"><img border="0" src="https://blogger.googleusercontent.com/img/b/R29vZ2xl/AVvXsEhm-HVxRc2MEIpbNis7OoTBLqfWnsiy-iNd3I6EaAcc3wlyVzDwxnpLmoTIA0gi8I6yxVHgSkWxFXndvWivWDMn6YFAmF6TS-7hc2yGIhZRWunNfynVlwspvfYeUH5-zYzV7f0VGimOcDI7W3YtUEUYHZQU0r8kY_NprDVTamIOD600PBohHIb3cxv4t-bl/s16000/interrupt_new_01.png"></a></td></tr>
<tr><td>Interrupt handling during NOP on NXT</td></tr>
</tbody></table>
<p><br></p>
<div><p>On cycle 57, the INTR line off the CPU was asserted. However, this has no immediate effect until the next cycle rising edge, where an internal latch is set (See Ken's excellent article on <a href="https://www.righto.com/2023/02/8086-interrupt.html" target="_blank">interrupts on the 8086</a>)</p></div>
<p><br></p>
<p>This latch is applied with other logic to result in an internal INTR signal, if interrupts are not disabled*. This signal is then checked when the CPU is about to fetch the next instruction byte. </p>
<p><br></p>
<blockquote><p>*Interrupts can be disabled via the Interrupt CPU flag, or temporarily after certain instructions such as STI or access to segment registers. </p></blockquote>
<p><br></p>
<p>Intel's documentation is not terribly specific about when interrupts are checked, simply stating that it is done at the end of an instruction.  Experimentally, we can show that the presence of an NXT flag* can affect interrupt timings.  If there is a byte available in the queue, the NXT flag on the microcode instruction will instruct the BIU to read it and prepare it as the next instruction. Before this is done, however, the internal interrupt signal is checked. If an interrupt is to be processed, the byte is not read.</p>
<p><br></p>
<blockquote><p>*The NXT flag is a signal to the CPU from the EU that the next microcode instruction to be executed is the last microcode instruction.  This allows the next instruction byte to be read from the queue, if available, and the process of preparing the next instruction to commence one cycle earlier than normal.  This enables a one-cycle execution pipeline, where the last microcode instruction executes on the first cycle of the new instruction - which would have otherwise left the EU's microcode execution unit idle. </p></blockquote>
<p>The <b>NOP </b>example above shows the effect of NXT. The next cycle is spent preparing the interrupt microcode routine, and then on cycle 60 the microcode executes a request for an interrupt acknowledge cycle. This enters the Tr state as normal. </p>
<p><br></p>
<p>Let's look at an example where the INTR line is asserted one cycle later in the NOP instruction.</p>
<p><br></p>
<table cellspacing="0" align="center" cellpadding="0"><tbody>
<tr><td><a href="https://blogger.googleusercontent.com/img/b/R29vZ2xl/AVvXsEhl_L-RdozDmUvxXPKXwHK_zKKHZqZmUoeakLC7_ptTk-Ert0f-_SvImjfwURK4iel3p7kfoDqVSKe2hijNhZ3MdFnu0l-Gk9Tf3w6gkMwrtuA6q4WcqaBqmzvbAIH4OUGH5pzJIzhrMLTeGE4BicwfWlpzzFPtQ3CIC7zy_kcSqo_rjj5TMG9TtgXXo3dH/s796/interrupt_new_rni_01.png"><img border="0" src="https://blogger.googleusercontent.com/img/b/R29vZ2xl/AVvXsEhl_L-RdozDmUvxXPKXwHK_zKKHZqZmUoeakLC7_ptTk-Ert0f-_SvImjfwURK4iel3p7kfoDqVSKe2hijNhZ3MdFnu0l-Gk9Tf3w6gkMwrtuA6q4WcqaBqmzvbAIH4OUGH5pzJIzhrMLTeGE4BicwfWlpzzFPtQ3CIC7zy_kcSqo_rjj5TMG9TtgXXo3dH/s16000/interrupt_new_rni_01.png"></a></td></tr>
<tr><td>Interrupt handling during NOP on RNI</td></tr>
</tbody></table>
<p>Here the INTR line is physically asserted on cycle 15, and latched on the leading edge of 16, in time for the RNI flagged microcode instruction to execute.  Like NXT, RNI will attempt to process the next instruction byte, but checks the interrupt status first.  Finding an active, enabled interrupt signal, we prepare to execute the interrupt microcode on cycle 17.  Note that the BIU is not really aware of this, and so on T2, it starts a new code fetch bus cycle.</p>
<p><br></p>
<p>The request for an interrupt acknowledge bus cycle on cycle 18 then follows standard prefetch abort logic. On T4, the abort check occurs, and only then is the interrupt acknowledge bus cycle begun.</p>
<p><br></p>
<p>We end up with completely different interrupt timings based off a single cycle difference in when INTR is asserted, due to the effect of the NXT flag.</p>
<p><br></p>
<p>An instruction such as <b>INTO </b>(when the overflow flag is not set) has identical cycle timings to <b>NOP</b>, however does not have any NXT flagged microcode instructions. In this case, we observe no difference between INTR being asserted on any cycle of INTO prior to RNI.</p>
<p><br></p>
<h2>Halt Timings</h2>
<p>We've addressed pretty much every previously odd bus scenario on the 8088, leaving us to address the timings around the <b>HLT </b>instruction and its related halt state.  Unfortunately, our new model does not significantly clarify the situation here.  <b>HLT </b>is not a microcoded instruction; it is implemented with a series of specialized logic gates.</p>
<p><br></p>
<p><br></p>
<div>
<p>The <b>HLT</b> instruction is therefore a state machine that proceeds independently of the BIU logic, which continues independently. This has some interesting ramifications. </p>
<p><br></p>
<p>First, it is possible to fully prevent the one-cycle HALT bus state from occurring if an interrupt is already pending, or INTR is raised high during the first clock of the <b>HLT </b>instruction: </p>
<table cellspacing="0" align="center" cellpadding="0"><tbody>
<tr><td><a href="https://blogger.googleusercontent.com/img/b/R29vZ2xl/AVvXsEgKUaW013xSB0GOENP_R1OKK4HzLQU_wprmlt3kSZ__Uxn9cWACAD74xlesUG5cezphu5qPxyOGQn9oimfmqvTMJqPS-y1ClMbFCIU5I_8Q4Q4LkYYk1MRnU0-rWdimYRsPPlhRIe9kBBbaaazLYHyQ3peuf67d-MeqZcdYXtaiLCmt7BWsjnA0f-_LoD2j/s794/halt_prevented_02.png"><img border="0" src="https://blogger.googleusercontent.com/img/b/R29vZ2xl/AVvXsEgKUaW013xSB0GOENP_R1OKK4HzLQU_wprmlt3kSZ__Uxn9cWACAD74xlesUG5cezphu5qPxyOGQn9oimfmqvTMJqPS-y1ClMbFCIU5I_8Q4Q4LkYYk1MRnU0-rWdimYRsPPlhRIe9kBBbaaazLYHyQ3peuf67d-MeqZcdYXtaiLCmt7BWsjnA0f-_LoD2j/s16000/halt_prevented_02.png"></a></td></tr>
<tr><td>HALT bus state suppressed by early INTR</td></tr>
</tbody></table>
<p><br></p>
<p>The situation is different if INTR is asserted even a single cycle later:</p>
<table cellspacing="0" align="center" cellpadding="0"><tbody>
<tr><td><a href="https://blogger.googleusercontent.com/img/b/R29vZ2xl/AVvXsEg3t7IozlcMbc_9OTcc3lvAsP2Y86-PuRstpfphWqAtFAKQb42AlqxGP7CFWEFH7D-RO-F1-3D-yKFCtHxNqNHFr9V3SBmzWERPd_gesMhjwGc0CZmDOJUzH8srEtksrExZbtSVxT_erq0GKaglSFDCFt6v1LEIJ8YG5UGOBF6urIpw183HsTy8UPgw4qao/s790/halt_resume_fast_01.png"><img border="0" src="https://blogger.googleusercontent.com/img/b/R29vZ2xl/AVvXsEg3t7IozlcMbc_9OTcc3lvAsP2Y86-PuRstpfphWqAtFAKQb42AlqxGP7CFWEFH7D-RO-F1-3D-yKFCtHxNqNHFr9V3SBmzWERPd_gesMhjwGc0CZmDOJUzH8srEtksrExZbtSVxT_erq0GKaglSFDCFt6v1LEIJ8YG5UGOBF6urIpw183HsTy8UPgw4qao/s16000/halt_resume_fast_01.png"></a></td></tr>
<tr><td>Fastest HALT resume timing</td></tr>
</tbody></table>
<p><br></p>
<div>
<p>In this case we do see the HALT state; however we resume from halt a mere two cycles later. </p>
<p><br></p>
</div>
</div>
<p>When resuming from halted state, the interrupt logic may occur more or less as we have already established. The INTR line goes high, is latched on the next rising edge transition of the CPU clock, the microcode routine for interrupt is loaded, and a bus request for an interrupt acknowledge cycle is made. This enters the Tr state, and we begin T1 of an interrupt acknowledge bus cycle exactly when anticipated:</p>
<p><br></p>
<table cellspacing="0" align="center" cellpadding="0"><tbody>
<tr><td><a href="https://blogger.googleusercontent.com/img/b/R29vZ2xl/AVvXsEhvGaCjkvjOOnhLcrbhJ-wwNJ7mex0Wy4_1c-c6Y9UoSiUJWA5Jp5X1o_NZ-Ja_PCFL4jEyAPV5QNoTlvyRVGQ392JOY3KdVZSLZ6D9chqmu4gJ3qh22jxaI8LND0md0iA14OtqBn7FlO4kBmruHytqrDa7Jfk1AJGztVuooDg9rtbhevsxVktIp3DMjYDo/s792/normal_halt_resume_01.png"><img border="0" src="https://blogger.googleusercontent.com/img/b/R29vZ2xl/AVvXsEhvGaCjkvjOOnhLcrbhJ-wwNJ7mex0Wy4_1c-c6Y9UoSiUJWA5Jp5X1o_NZ-Ja_PCFL4jEyAPV5QNoTlvyRVGQ392JOY3KdVZSLZ6D9chqmu4gJ3qh22jxaI8LND0md0iA14OtqBn7FlO4kBmruHytqrDa7Jfk1AJGztVuooDg9rtbhevsxVktIp3DMjYDo/s16000/normal_halt_resume_01.png"></a></td></tr>
<tr><td>"Normal" halt resume timings</td></tr>
</tbody></table>
<p><br></p>
<p>However, it is often observed that T1 of the interrupt acknowledge bus cycle is delayed by one cycle relative to INTR.  This phenomenon has not yet been conclusively explained.</p>
<p><br></p>
<table cellspacing="0" align="center" cellpadding="0"><tbody>
<tr><td><a href="https://blogger.googleusercontent.com/img/b/R29vZ2xl/AVvXsEgZDIPwgb77dMAEdTzSDstPMaHaTjp_f47oNXKj2_3aRWiJOlIzc48AoBLZFnZgYnoObOw2t0dBlYaLvsnRFh1ZH8-BWEyvelR98Dmp8QL4MTp-a7HFuDlMPGdug90CVb1Sw4B3RM2cOFhiaN4ahGXQ18CpGzv4q4CZoTvB23E-YLgqGrdB-BxDKXLl4Iog/s792/delayed_halt_resume_01.png"><img border="0" src="https://blogger.googleusercontent.com/img/b/R29vZ2xl/AVvXsEgZDIPwgb77dMAEdTzSDstPMaHaTjp_f47oNXKj2_3aRWiJOlIzc48AoBLZFnZgYnoObOw2t0dBlYaLvsnRFh1ZH8-BWEyvelR98Dmp8QL4MTp-a7HFuDlMPGdug90CVb1Sw4B3RM2cOFhiaN4ahGXQ18CpGzv4q4CZoTvB23E-YLgqGrdB-BxDKXLl4Iog/s16000/delayed_halt_resume_01.png"></a></td></tr>
<tr><td>"Delayed" halt resume timings</td></tr>
</tbody></table>
<p><br></p>
<p>I propose the following theory: If we follow the logic Ken describes in his blog, we read that <b>HLT </b>logic<b> </b>waits for T2, whereupon it disables prefetching.  However, depending on current bus activity, T2 may occur before HLT can execute this check, and thus a prefetch bus cycle is started before HLT can disable prefetching.  It is this scenario that appears to cause a one cycle delay when resuming from a halted state.  This could simply be the result of the incomplete fetch cycle, but its possible that the states of the internal latches that drive the halt condition could be responsible.</p>
<h2>Conclusion</h2>
<p>After attempting to piece the puzzle of the 8088 together over the past year, I'm extremely satisfied to finally have a simple, consistent description of the chip's bus timings and prefetch behavior. I can only apologize for publishing an incomplete model in my previous article - I hope it did not lead too many people astray. The key to this simplified model just required accepting something contrary to accepted wisdom. It is often difficult to look past our initial assumptions: in this case, that a bus cycle on the 8088 must be four cycles long.</p>
<p><br></p>
<div><p>A big thank you, again, to Ken Shirriff and <a href="https://www.righto.com/" target="_blank">his excellent blog</a>. If you aren't following it, you're missing out!</p></div>
<p><br></p>
<p><br></p>
<p><br></p>
<p>.</p>
</div>
</div></div></div></div><br>
<br>
<br>
<br>
<br>
<br>
via Adventures in PC Emulation <br>
<br>
February 14, 2024 at 11:09PM
